src/object/deduplicate_key.c

```diff
@@ -0,0 +1,497 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * deduplicate_key.h - Support duplicate key index
+ */
+
+#ident "$Id$"
+
+#include "config.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "dbtype.h"
+#if defined(SERVER_MODE) || defined(SA_MODE)
+#include "object_representation.h"
+#include "object_primitive.h"
+#include "object_representation_sr.h"
+#endif
+
+#include "deduplicate_key.h"
+
+#if defined(SUPPORT_DEDUPLICATE_KEY_MODE)
+
+// The higher the level, the more severe the deduplicate.
+#define CALC_MOD_VALUE_FROM_LEVEL(lv)   (1 << (lv))
+
+static char dk_reserved_deduplicate_key_attr_name[COUNT_OF_DEDUPLICATE_KEY_LEVEL][DEDUPLICATE_KEY_ATTR_NAME_LEN];
+
+static DB_DOMAIN *
+get_reserved_index_attr_domain_type (int level)
+{
+  if (level == DEDUPLICATE_KEY_LEVEL_NONE)
+    {
+      return &tp_Integer_domain;
+    }
+
+  return (CALC_MOD_VALUE_FROM_LEVEL (level) > SHRT_MAX) ? &tp_Integer_domain : &tp_Short_domain;
+}
+
+//=============================================================================
+#if defined(SERVER_MODE) || defined(SA_MODE)
+static OR_ATTRIBUTE st_or_atts[COUNT_OF_DEDUPLICATE_KEY_LEVEL];
+static bool st_or_atts_init = false;
+
+static void
+dk_or_attribute_initialized ()
+{
+  int att_id;
+  int level, idx;
+
+  for (level = DEDUPLICATE_KEY_LEVEL_MIN; level <= DEDUPLICATE_KEY_LEVEL_MAX; level++)
+    {
+      att_id = MK_DEDUPLICATE_KEY_ATTR_ID (level);
+      idx = LEVEL_2_IDX (level);
+      st_or_atts[idx].id = att_id;
+      st_or_atts[idx].domain = get_reserved_index_attr_domain_type (level);
+      st_or_atts[idx].type = st_or_atts[idx].domain->type->id;
+    }
+
+#ifndef NDEBUG
```

Comment:
need comment for the reason of debugging

Reply:
Check whether the macro that creates or extracts the attribute name and ID for the deduplicate_key for the specified level operates normally.

I will add a comment.